pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,545 @@
+    h10002: Sequence[complex] | complex = None
+    # sextupole geometric rdts
+    h21000: Sequence[complex] | complex = None
+    h30000: Sequence[complex] | complex = None
+    h10110: Sequence[complex] | complex = None
+    h10020: Sequence[complex] | complex = None
+    h10200: Sequence[complex] | complex = None
+    # octupole geometric rdts
+    h22000: Sequence[complex] | complex = None
+    h11110: Sequence[complex] | complex = None
+    h00220: Sequence[complex] | complex = None
+    h31000: Sequence[complex] | complex = None
+    h40000: Sequence[complex] | complex = None
+    h20110: Sequence[complex] | complex = None
+    h11200: Sequence[complex] | complex = None
+    h20020: Sequence[complex] | complex = None
+    h20200: Sequence[complex] | complex = None
+    h00310: Sequence[complex] | complex = None
+    h00400: Sequence[complex] | complex = None
+    # Detuning
+    dnux_dJx: Sequence[float] | float = None
+    dnux_dJy: Sequence[float] | float = None
+    dnuy_dJy: Sequence[float] | float = None
+
+    def __getattr__(self, item):
+        return asdict(self)[item]
+
+    def __getitem__(self, item):
+        return asdict(self)[item]
+
+
+def _get_polynom(elem, attr, index):
+    try:
+        val = getattr(elem, attr)[index]
+        return val
+    except (IndexError, AttributeError):
+        return 0
+
+
+def _computedrivingterms(
+    s,
+    betax,
+    betay,
+    phix,
+    phiy,
+    etax,
+    a2l,
+    b2l,
+    b3l,
+    b4l,
+    tune,
+    rdttype,
+    nperiods,
+    refpt,
+    second_order,
+):
+    """
+    Original implementation from ELEGANT
+    Based on J.Bengtsson, SLS Note 9 / 97, March 7, 1997, with corrections per W.Guo (NSLS)
+    Revised to follow C.X.Wang AOP - TN - 2009 - 020 for second - order terms
+    """
+    periodicfactor = numpy.ones((9, 9))
+    rdts = {}
+
+    def pf(i, j):
+        return periodicfactor[4 + i][4 + j]
+
+    if nperiods != 1:
+        for i in range(9):
+            for j in range(9):
+                a1 = numpy.pi * 2 * (tune[0] * (i - 4) + tune[1] * (j - 4))
+                a2 = a1 / nperiods
+                periodicfactor[i][j] = (numpy.exp(1j * a1) - 1.0) / (
+                    numpy.exp(1j * a2) - 1.0
+                )
+
+    rbetax = numpy.sqrt(betax)
+    rbetay = numpy.sqrt(betay)
+    px = numpy.exp(1j * phix)
+    py = numpy.exp(1j * phiy)
+
+    rdts["refpts"] = refpt
+    mask_a2l = numpy.absolute(a2l) > 1.0e-6
+    mask_b2l = numpy.absolute(b2l) > 1.0e-6
+    mask_b3l = numpy.absolute(b3l) > 1.0e-6
+    mask_b4l = numpy.absolute(b4l) > 1.0e-6
+
+    if (RDTType.FOCUSING in rdttype) or (RDTType.ALL in rdttype):
+        b2lm = b2l[mask_b2l]
+        betaxm = betax[mask_b2l]
+        betaym = betay[mask_b2l]
+        pxm = px[mask_b2l]
+        pym = py[mask_b2l]
+        rdts["h20000"] = sum((b2lm / 8) * betaxm * pxm * pxm * pf(2, 0))
+        rdts["h00200"] = sum((b2lm / 8) * betaym * pym * pym * pf(0, 2))
```

Comment:
I have derived `h00200` and it should have a negative sign in AT.
Here below the explanation.
I believe Eq(94) in Bengtsson has a typo, there should be a minus in front of $b_2$ to be consistent with Eq(10).
This typo has no effect in third order RDTs, therefore, no other changes are needed.
I show part of the derivation in case you would like to check it.
$V_i = \frac{-b_2}{2(1+\delta)}(x^2-y^2)=\frac{-b_2}{2}(1-\delta)(x^2-y^2)$
Ignoring the terms in $\delta$
$V_i=\frac{-b_2}{2}x^2+\frac{b_2}{2}y^2$
Ignoring the terms in $x$, and using Eq(72)
$\hat V_i = R_{o\rightarrow i}A_iV_i = \frac{b_2}{2}\beta_yR_{o \rightarrow i} (\frac{h_y^+ + h_y^-}{2})^2$,
 which is the equivalent to Eqs(89,90,91) for the $y$ plane.
We expand  the polynom on the resonant base and write the term with only $h_y^+$
$\hat V_i=\frac{b_2}{2}\beta_y \{{\frac{1}{4}(h_y^+ + h_y^- )^2}} = \frac{b_2}{2}\beta_y \frac{1}{4} ((h_y^+)^2 e^{i2\mu_y} + \cdots )$
Integrating as in Eq(63)
$h_{00200}=\frac{1}{8} \beta_y b_2 L  e^{i2\mu_y}$
Changing the sign to match the AT magnetic field definition(Bengtsson has minus in Eq(10)).
$h_{00200,AT}=-\frac{1}{8} \beta_y b_2L  e^{i2\mu_y}$

Reply:
Same holds for h20000 in this case right?

Reply:
In fact no... this was a serious bug... thanks!